bin/i18n/sync-codeorg-in.rb

```diff
@@ -6,173 +6,137 @@
 require 'json'
-require 'yaml'
-require 'tempfile'
 
 require_relative 'i18n_script_utils'
 
+I18N_SOURCE_DIR = "i18n/locales/source"
+
 def sync_in
   localize_level_content
   localize_block_content
+  puts "Copying source files"
   run_bash_script "bin/i18n-codeorg/in.sh"
   redact_level_content
   redact_block_content
 end
 
-def copy_to_yml(label, data)
-  File.open("dashboard/config/locales/#{label}.en.yml", "w+") do |f|
-    f.write(to_crowdin_yaml({"en" => {"data" => {label => data}}}))
-  end
-end
+def get_i18n_strings(level)
+  i18n_strings = {}
+
+  if level.is_a?(DSLDefined)
+    text = level.dsl_text
+    i18n_strings["dsl"] = level.class.dsl_class.parse(text, '')[1] if text
+  elsif level.is_a?(Level)
+    %w(
+      display_name
+      short_instructions
+      long_instructions
+      failure_message_overrides
+    ).each do |prop|
+      i18n_strings[prop] = level.try(prop)
+    end
 
-# sanitize a string before uploading to crowdin. Currently only performs
-# CRLF -> LF conversion, but could be extended to do more
-def sanitize(string)
-  return string.gsub(/\r(\n)?/, "\n")
-end
+    # authored_hints
+    if level.authored_hints
+      authored_hints = JSON.parse(level.authored_hints)
+      i18n_strings['authored_hints'] = Hash.new unless authored_hints.empty?
+      authored_hints.each do |hint|
+        i18n_strings['authored_hints'][hint['hint_id']] = hint['hint_markdown']
+      end
+    end
 
-def redact_translated_data(path, plugins = nil)
-  source = "i18n/locales/source/#{path}"
-  backup = "i18n/locales/original/#{path}"
-  FileUtils.mkdir_p(File.dirname(backup))
-  FileUtils.cp(source, backup)
-  redact(source, source, plugins)
-end
+    # callouts
+    if level.callout_json
+      callouts = JSON.parse(level.callout_json)
+      i18n_strings['callouts'] = Hash.new unless callouts.empty?
+      callouts.each do |callout|
+        i18n_strings['callouts'][callout['localization_key']] = callout['callout_text']
+      end
+    end
 
-def redact_block_content
-  redact_translated_data('dashboard/blocks.yml', 'blockfield')
-end
+    level_xml = Nokogiri::XML(level.to_xml, &:noblanks)
+    blocks = level_xml.xpath('//blocks').first
+    if blocks
+      ## Categories
+      block_categories = blocks.xpath('//category')
+      i18n_strings['block_categories'] = Hash.new unless block_categories.empty?
+      block_categories.each do |category|
+        name = category.attr('name')
+        i18n_strings['block_categories'][name] = name if name
+      end
+
+      ## Function Names
+      functions = blocks.xpath("//block[@type=\"procedures_defnoreturn\"]")
+      i18n_strings['function_names'] = Hash.new unless functions.empty?
+      functions.each do |function|
+        name = function.at_xpath('./title[@name="NAME"]')
+        i18n_strings['function_names'][name.content] = name.content if name
+      end
+    end
+  end
+
+  i18n_strings["contained levels"] = level.contained_levels.map do |contained_level|
```

Comment:
i'm building a new BubbleChoice level type somewhat similar to contained levels, but uses "sublevels" (ex: `level.sublevels` returns an array of levels). it's still in pre-pilot phase, so we don't need to work through this right now, but we will have to incorporate this level type's sublevels in here at some point. i'm open to suggestions for whatever is easiest for you. i'll follow up with you about this, but i wanted to give you a heads up

Reply:
Oh, thanks for letting me know! I'd love to hear all about what's happening/planned with that.

For some further context for you, I'm actually very interested in getting rid of contained levels entirely and finding a way to achieve the same functionality with just a single level, because the dispersal of relevant display data between multiple levels has caused us a lot of issues (the need for this PR being one of them). I'd really like to get a sense of what the plan is for data ownership with BubbleChoice levels and sublevels, and make sure that we're all thinking about data the same way.

Reply:
that would be great! here are some resources if you want to poke around: [PR that models BubbleChoice](https://github.com/code-dot-org/code-dot-org/pull/28756) (includes the spec, tech spec, and some jira info)

i scheduled some time for us to talk on tuesday, too!